File: Programmers/Algorithm_Kit/10_graph/001.cpp
#include <string>
#include <vector>
#include <queue>
#include <iostream>
using namespace std;

int visited[20001], max_dist;
vector<int> adj[20001];

void bfs(int here) {
    visited[here] = 1;

    queue<int> q;
    q.push(here);

    while (!q.empty()) {
        int cur = q.front(); q.pop();

        for (int next : adj[cur]) {
            if (!visited[next]) {
                visited[next] = visited[cur] + 1;
                q.push(next);

                max_dist = visited[next];
            }
        }
    }

    return;
}

int solution(int n, vector<vector<int>> edge) {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL); cout.tie(NULL);

    for (int i = 0; i < edge.size(); i++) {
        adj[edge[i][0]].push_back(edge[i][1]);
        adj[edge[i][1]].push_back(edge[i][0]);
    }

    bfs(1);

    int answer = 0;

    for (int i = 1; i <= n; i++) {
        if (visited[i] == max_dist) {
            answer++;
        }
    }

    return answer;
}

File: programmers/algorithmkit/10_graph/001.java
import java.util.*;

class Solution {
    int maxDist;
    int[] visited = new int[20001];
    List<Integer>[] adj = new ArrayList[20001];

    public int solution(int n, int[][] edge) {
        int answer = 0;

        // init
        for (int i = 1; i <= n; i++) {
            adj[i] = new ArrayList<>();
        }

        for (int[] e : edge) {
            adj[e[0]].add(e[1]);
            adj[e[1]].add(e[0]);
        }

        bfs(1);

        for (int i = 1; i <= n; i++) {
            if (visited[i] == maxDist) {
                answer++;
            }
        }

        return answer;
    }

    void bfs(int start) {
        visited[start] = 1;
        maxDist = visited[start];

        Queue<Integer> q = new ArrayDeque<>();
        q.offer(start);

        while (!q.isEmpty()) {
            int cur = q.poll();

            for (int next : adj[cur]) {
                if (visited[next] == 0) {
                    visited[next] = visited[cur] + 1;
                    q.offer(next);

                    maxDist = Math.max(maxDist, visited[next]);
                }
            }
        }
    }
}
